x/reports/internal/keeper/keeper.go

```diff
@@ -0,0 +1,100 @@
+package keeper
+
+import (
+	"bytes"
+
+	"github.com/cosmos/cosmos-sdk/codec"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/desmos-labs/desmos/x/posts"
+	"github.com/desmos-labs/desmos/x/reports/internal/types"
+	"github.com/desmos-labs/desmos/x/reports/internal/types/models"
+)
+
+// Keeper maintains the link to data storage and exposes getter/setter methods for the various parts of the state machine
+type Keeper struct {
+	PostKeeper posts.Keeper // post's keeper to perform checks on the postIDs
+	StoreKey   sdk.StoreKey // Unexposed key to access store from sdk.Context
+	Cdc        *codec.Codec // The wire codec for binary encoding/decoding.
+}
+
+// NewKeeper creates new instances of the reports Keeper
+func NewKeeper(pk posts.Keeper, cdc *codec.Codec, storeKey sdk.StoreKey) Keeper {
+	return Keeper{
+		PostKeeper: pk,
+		StoreKey:   storeKey,
+		Cdc:        cdc,
+	}
+}
+
+// SaveReport allows to save the given reports inside the current context.
+// It assumes that the given reports has already been validated.
+// If the same reports has already been inserted, nothing will be changed.
+func (k Keeper) SaveReport(ctx sdk.Context, postID posts.PostID, report types.Report) bool {
+	store := ctx.KVStore(k.StoreKey)
+	key := models.ReportStoreKey(postID)
+	// Get the list of reports related to the given postID
+	var reports models.Reports
+	k.Cdc.MustUnmarshalBinaryBare(store.Get(key), &reports)
+
+	// try to append the given reports
+	reports, appended := reports.AppendIfMissing(report)
+	if appended {
+		store.Set(key, k.Cdc.MustMarshalBinaryBare(&reports))
+	}
+
+	return appended
+}
+
+// RegisterReportsTypes allows to save the given reportType inside the chain state.
+func (k Keeper) RegisterReportsTypes(ctx sdk.Context, reportType types.ReportType) bool {
```

Comment:
I really don't know how to feel about the pre-registered report types and the fact that someone might not be able to create a report if the type isn't registered. 

I'm thinking about the possibility that some dApps might want to use custom report types that are more specific, for example Uber when you report your driver or something else. Having pre-registered and a fixed amount of report types could lead to this not being possible. 

For this reason, I think it's better to get rid of the pore-registered report type, de-facto allowing dApps and users to specify any report type they want when creating them. This should allow for a more easy usage of the protocol overall.

Reply:
While I'm implementing the pre-registered type lists I was referring to your statement in the linked issue:
>To be consistent across different applications, we should define a list of supported Types (probably inside the genesis) so that custom applications can have a list of them.

I'm fine with removing the pre-registered list cause it would lead to less checks during the report's saving but doing this we will lose the consistency that you are talking about in the  issue. If that's not important we could remove it w/o problems.

Reply:
@bragaz Yeah you're right, that was my initial idea. However, on a comment later on I asked to leave out this point while implementing the feature (see [here](https://github.com/desmos-labs/desmos/issues/50#issuecomment-628397322)). 

I think it's better to remove this all together to have more freedom and less checks so please proceed with the removal of pre-registered types.